src/main/java/com/gmail/woosay333/onlinebookstore/service/shopping/cart/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,102 @@
+
+import com.gmail.woosay333.onlinebookstore.dto.shopping.cart.CartItemDto;
+import com.gmail.woosay333.onlinebookstore.dto.shopping.cart.CartItemUpdateRequestDto;
+import com.gmail.woosay333.onlinebookstore.dto.shopping.cart.CreateCartItemRequestDto;
+import com.gmail.woosay333.onlinebookstore.dto.shopping.cart.ShoppingCartDto;
+import com.gmail.woosay333.onlinebookstore.entity.Book;
+import com.gmail.woosay333.onlinebookstore.entity.CartItem;
+import com.gmail.woosay333.onlinebookstore.entity.ShoppingCart;
+import com.gmail.woosay333.onlinebookstore.entity.User;
+import com.gmail.woosay333.onlinebookstore.exception.EntityNotFoundException;
+import com.gmail.woosay333.onlinebookstore.mapper.CartItemMapper;
+import com.gmail.woosay333.onlinebookstore.mapper.ShoppingCartMapper;
+import com.gmail.woosay333.onlinebookstore.repository.book.BookRepository;
+import com.gmail.woosay333.onlinebookstore.repository.shopping.cart.CartItemRepository;
+import com.gmail.woosay333.onlinebookstore.repository.shopping.cart.ShoppingCartRepository;
+import com.gmail.woosay333.onlinebookstore.service.shopping.cart.ShoppingCartService;
+import com.gmail.woosay333.onlinebookstore.service.user.UserService;
+import jakarta.transaction.Transactional;
+import java.util.Optional;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final CartItemRepository cartItemRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemMapper cartItemMapper;
+    private final UserService userService;
+    private final BookRepository bookRepository;
+
+    @Override
+    public ShoppingCartDto getShoppingCart() {
+        User user = userService.getAuthenticatedUser();
+        return shoppingCartMapper.toDto(shoppingCartRepository.findByUserId(user.getId())
+                .orElseGet(() -> createNewShoppingCart(user)));
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartDto addToCart(CreateCartItemRequestDto cartItemRequestDto) {
+        User user = userService.getAuthenticatedUser();
+        Book book = bookRepository.findById(cartItemRequestDto.getBookId())
+                .orElseThrow(() -> new EntityNotFoundException(
+                        String.format("Book with id: %d not found",
+                                cartItemRequestDto.getBookId())));
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(user.getId())
+                .orElseGet(() -> createNewShoppingCart(user));
+        CartItem cartItem = createCartItem(book, shoppingCart, cartItemRequestDto.getQuantity());
+        cartItemRepository.save(cartItem);
+        shoppingCart.getCartItems().add(cartItem);
+        return shoppingCartMapper.toDto(shoppingCart);
```

Comment:
I don`t see the logic that saves shopping cart after it was updated with new cartItem

Reply:
According to the logic of my application, the creation of a shopping cart occurs at the moment the user adds an item to the basket. Did I understand you correctly that I need to move the logic of creating a shopping cart to the moment the user registers in the system?